src/visitors/sympy_replace_solutions_visitor.hpp

```diff
@@ -0,0 +1,504 @@
+/*************************************************************************
+ * Copyright (C) 2018-2021 Blue Brain Project
+ *
+ * This file is part of NMODL distributed under the terms of the GNU
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+
+#pragma once
+
+/**
+ * \file
+ * \brief \copybrief nmodl::visitor::SympyReplaceSolutionsVisitor
+ */
+
+#include "visitors/ast_visitor.hpp"
+
+#include <algorithm>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace nmodl {
+namespace visitor {
+
+/**
+ * @addtogroup visitor_classes
+ * @{
+ */
+
+
+/**
+ * \class SympyReplaceSolutionsVisitor
+ * \brief Replace statements in \p node with pre_solve_statements, tmp_statements, and solutions
+ *
+ * The goal is to replace statements with \ref solutions_ in place. In this way we can allow (to
+ * some extent) the use of control flow blocks and assignments. \ref pre_solve_statements are added
+ * in front of the replaced statements in case their variable needs updating. \ref
+ StatementDispenser
+ * keeps track of what needs updating. Let's start with some nomenclature:
+ *
+ * - statement: a line in the .mod file. It can be a diff_eq_expression, binary_expression, or
+ * linEquation
+ * - old_Statement: line in the staementBlock that must be replaced with the solution
+ * - solution/ new_statement: a nmodl-statement (always binary expression) provided by sympy that
+ * assigns a variable
+ * - pre_solve_Statements: statements that update the variables (i.e. x = old_x)
+ * - tmp_solutions: assignment of temporary variables in the solution generated by sympy in case
+ * --cse. (i.e. \f tmp = f
+ * (...) \f
+ *
+ * We employ a multi-step approach:
+ *
+ * - try to replace the old_statements (not binary_expressions) and in "assignment form: \f x =
+ * f(...) \f" with the corresponding solution matching by variable (i.e. x in \f x = f(...) \f)
+ * - try to replace the old_Statements with a greedy approach. When we find a
+ * diff_eq_expression/linEquation that needs replacing we take the next solution that was not yet
+ * used
+ * - add all the remaining solutions at the end
+ *
+ * Let's finish with an example (that are usually better than blabbling around).
+ *
+ * Imagine we have this derivative block in the ast (before SympyReplaceSolutionsVisitor passes):
+ *
+ * \code
```

Comment:
```suggestion
 * \code{.mod}
```
the reason for this is that with `.mod`, doxygen does NMODL syntax highlighting in the docs!

Reply:
aaaaah. I did not understand